docs/configuration.md
# Configuration

`PixelFS` allows you to modify certain configurations, and all configuration details are stored in the `$HOME/.pixelfs/config.toml` file.

::: tip
If you are running `PixelFS` for the first time, a `config.toml` file will be automatically created.
:::

## Configuration Parameters

| Parameter Name             | Type     |         Default Value         | Description                                         |
|----------------------------|----------|:-----------------------------:|-----------------------------------------------------|
| endpoint                   | string   |   `https://www.pixelfs.io`    | Server address                                      |
| token                      | string   |               -               | User TOKEN information                              |
| pwd                        | string   |               -               | Path information used with `pixelfs cd`             |
| debug                      | bool     |            `false`            | Enable debug logs                                   |
| daemon.listen              | string   |        `0.0.0.0:15233`        | Address and port for the `daemon` service           |
| ffmpeg.cache.expire        | int      |            `86400`            | `ffmpeg` cache expiration time (seconds)            |
| ffmpeg.cache.path          | string   | `$HOME/.pixelfs/cache/ffmpeg` | `ffmpeg` cache path                                 |
| webdav.listen              | string   |        `0.0.0.0:5233`         | Address and port for the `webdav` service           |
| webdav.cache.expire        | int      |            `86400`            | `webdav` cache expiration time (seconds)            |
| webdav.cache.path          | string   | `$HOME/.pixelfs/cache/webdav` | `webdav` cache path                                 |
| webdav.cors.credentials    | bool     |               -               | Support for CORS credentials                        |
| webdav.cors.allow_origin   | string[] |            `['*']`            | Allowed origins for CORS requests                   |
| webdav.cors.allow_headers  | string[] |            `['*']`            | Allowed headers for CORS requests                   |
| webdav.cors.allow_methods  | string[] |            `['*']`            | Allowed methods for CORS requests                   |
| webdav.cors.expose_headers | string[] |               -               | Response headers accessible by clients              |
| webdav.cors.max_age        | int      |               -               | Maximum cache time for preflight requests (seconds) |
| webdav.users.username      | string   |               -               | Username for the `webdav` service                   |
| webdav.users.password      | string   |               -               | Password for the `webdav` service                   |
| webdav.users.permissions   | string   |               -               | `webdav` user permissions, e.g., `CRUD`             |

## Configuration Example

```toml
endpoint = 'https://www.pixelfs.io'
pwd = '/'
token = 'auth-token'

[daemon]
listen = '0.0.0.0:15233'

[ffmpeg]
[ffmpeg.cache]
expire = 86400
path = '$HOME/.pixelfs/cache/ffmpeg'

[webdav]
listen = '0.0.0.0:5233'

[webdav.cache]
expire = 86400
path = '$HOME/.pixelfs/cache/webdav'

[webdav.cors]
allow_headers = ['*']
allow_methods = ['*']
allow_origin = ['*']

[[webdav.users]]
username = 'admin'
password = '123456'
permissions = 'none'

[[webdav.users.rules]]
path = '/0x29e3abdb587207dc4ac9c708670eefde717ef307'
permissions = 'CRUD'
```


docs/zh-hans/configuration.md
# 配置

`PixelFS` 允许你修改一些配置，所有配置信息都存放在 `$HOME/.pixelfs/config.toml` 文件中。

::: tip 注意
如果你是首次运行 `PixelFS`, 则会自动创建一个 `config.toml` 文件。
:::

## 参数配置说明

| 参数名                      | 类型      |              默认值              | 说明                               |
|----------------------------|----------|:-----------------------------:|------------------------------------|
| endpoint                   | string   |   `https://www.pixelfs.io`    | 服务端地址                           |
| token                      | string   |               -               | 用户 TOKEN 信息                     |
| pwd                        | string   |               -               | 使用 `pixelfs cd` 切换的路径信息     |
| debug                      | bool     |            `false`            | 是否开启调试日志                     |
| daemon.listen              | string   |        `0.0.0.0:15233`        | `daemon` 服务监听的地址和端口         |
| ffmpeg.cache.expire        | int      |            `86400`            | `ffmpeg` 缓存时间（秒）              |
| ffmpeg.cache.path          | string   | `$HOME/.pixelfs/cache/ffmpeg` | `ffmpeg` 缓存路径                    |
| webdav.listen              | string   |        `0.0.0.0:5233`         | `webdav` 服务监听的地址和端口         |
| webdav.cache.expire        | int      |            `86400`            | `webdav` 缓存时间（秒）              |
| webdav.cache.path          | string   | `$HOME/.pixelfs/cache/webdav` | `webdav` 缓存路径                    |
| webdav.cors.credentials    | bool     |               -               | 是否支持跨域请求凭据（CORS）          |
| webdav.cors.allow_origin   | string[] |            `['*']`            | 允许的跨域请求来源                   |
| webdav.cors.allow_headers  | string[] |            `['*']`            | 允许的跨域请求头                     |
| webdav.cors.allow_methods  | string[] |            `['*']`            | 允许的跨域请求方法                   |
| webdav.cors.expose_headers | string[] |               -               | 可被客户端访问的响应头                |
| webdav.cors.max_age        | int      |               -               | 预检请求的最大缓存时间（秒）           |
| webdav.users.username      | string   |               -               | `webdav` 服务用户名                  |
| webdav.users.password      | string   |               -               | `webdav` 服务密码                    |
| webdav.users.permissions   | string   |               -               | `webdav` 用户权限，例如：`CRUD`        |

## 示例配置

```toml
endpoint = 'https://www.pixelfs.io'
pwd = '/'
token = 'auth-token'

[daemon]
listen = '0.0.0.0:15233'

[ffmpeg]
[ffmpeg.cache]
expire = 86400
path = '$HOME/.pixelfs/cache/ffmpeg'

[webdav]
listen = '0.0.0.0:5233'

[webdav.cache]
expire = 86400
path = '$HOME/.pixelfs/cache/webdav'

[webdav.cors]
allow_headers = ['*']
allow_methods = ['*']
allow_origin = ['*']

[[webdav.users]]
username = 'admin'
password = '123456'
permissions = 'none'

[[webdav.users.rules]]
path = '/0x29e3abdb587207dc4ac9c708670eefde717ef307'
permissions = 'CRUD'
```


config/config.go
package config

import (
	"path/filepath"
	"strings"

	"github.com/mitchellh/mapstructure"
	"github.com/pixelfs/pixelfs/util"
	"github.com/spf13/viper"
)

type Config struct {
	Endpoint string `toml:"endpoint"`
	Token    string `toml:"token"`
	Pwd      string `toml:"pwd"`
	Debug    bool   `toml:"debug"`
	Daemon   Daemon `toml:"daemon"`
	FFmpeg   FFmpeg `toml:"ffmpeg"`
	Webdav   Webdav `toml:"webdav"`
}

type Daemon struct {
	Listen string `toml:"listen"`
}

type FFmpeg struct {
	Cache   Cache          `toml:"cache"`
	Options map[string]any `toml:"options"`
}

type Webdav struct {
	Listen string `toml:"listen"`
	Prefix string `toml:"prefix"`
	Cache  Cache  `toml:"cache"`
	CORS   CORS   `toml:"cors"`
	Users  []User `toml:"users"`
}

type Cache struct {
	Path   string `toml:"path"`
	Expire int    `toml:"expire"`
}

type CORS struct {
	Credentials   bool     `toml:"credentials"`
	AllowOrigin   []string `toml:"allow_origin"`
	AllowHeaders  []string `toml:"allow_headers"`
	AllowMethods  []string `toml:"allow_methods"`
	ExposeHeaders []string `toml:"expose_headers"`
	MaxAge        int      `toml:"max_age"`
}

func LoadConfig(path string, isFile bool) error {
	home, err := util.GetHomeDir()
	if err != nil {
		return err
	}

	if isFile {
		viper.SetConfigFile(path)
	} else {
		viper.SetConfigName("config")
		if path == "" {
			viper.AddConfigPath(home)
		} else {
			viper.AddConfigPath(path)
		}
	}

	viper.SetConfigType("toml")

	viper.SetDefault("endpoint", "https://www.pixelfs.io")
	viper.SetDefault("daemon.listen", "0.0.0.0:15233")

	// ffmpeg
	viper.SetDefault("ffmpeg.cache.path", filepath.Join(home, "cache", "ffmpeg"))
	viper.SetDefault("ffmpeg.cache.expire", 3600*24)

	// webdav
	viper.SetDefault("webdav.listen", "0.0.0.0:5233")
	viper.SetDefault("webdav.cache.path", filepath.Join(home, "cache", "webdav"))
	viper.SetDefault("webdav.cache.expire", 3600*24)
	viper.SetDefault("webdav.cors.allow_origin", []string{"*"})
	viper.SetDefault("webdav.cors.allow_headers", []string{"*"})
	viper.SetDefault("webdav.cors.allow_methods", []string{"*"})

	viper.SetEnvPrefix("pixelfs")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv()

	if err := viper.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			if err := viper.SafeWriteConfig(); err != nil {
				return err
			}
		}
	}

	return nil
}

func Set(key string, v any) error {
	viper.Set(key, v)
	return viper.WriteConfig()
}

func Remove(key string) error {
	switch viper.Get(key).(type) {
	case string:
		viper.Set(key, "")
	default:
		viper.Set(key, nil)
	}

	return viper.WriteConfig()
}

func GetConfig() (*Config, error) {
	cfg := &Config{}
	err := viper.Unmarshal(cfg, viper.DecodeHook(
		mapstructure.ComposeDecodeHookFunc(
			mapstructure.StringToTimeDurationHookFunc(),
			mapstructure.StringToSliceHookFunc(","),
			mapstructure.TextUnmarshallerHookFunc(),
		),
	))

	if err != nil {
		return nil, err
	}

	return cfg, nil
}
